OpenSim/Simulation/GeodesicWrap/GeodesicWrappingMath.cpp

```diff
@@ -0,0 +1,572 @@
+    const ImplicitSurfaceParams& s,
+    ImplicitSurfaceState& q,
+    double eps          = 1e-13,
+    std::size_t maxIter = 100)
+{
+    Mat3x6 jacobian;
+
+    double error     = INF;
+    size_t iteration = 0;
+    for (; iteration < maxIter; ++iteration) {
+        SimTK::Vec3 vec = calcSurfaceProjectionConstraint(s, q);
+
+        // TODO use Inf norm.
+        error = vec.norm();
+
+        if (error < eps) {
+            return iteration;
+        }
+
+        calcSurfaceProjectionConstraintJacobian(s, q, jacobian);
+
+        // TODO use other decomposition if needed.
+        Vec6 step = -jacobian.colPivHouseholderQr().solve(vec);
+        q += step;
+    }
+
+    // TODO use opensim's assert
+    throw std::runtime_error("Failed to project state to implicit surface");
+}
+
+std::pair<ImplicitGeodesicState, ImplicitGeodesicState> 
+ImplicitSurfaceParams::
+calcLocalGeodesic(
+        State qStart,
+        double length,
+        std::vector<SimTK::Vec3>& log) const
+{
+    size_t steps = _impl->integrationSteps;
+    double s   = 0.;
+    double ds  = length / static_cast<double>(steps);
+    ImplicitGeodesicState xStart(std::move(qStart));
+    ImplicitGeodesicState xEnd = xStart;
+
+    log.clear();
+    log.resize(steps);
+    log.push_back(xEnd.postion);
+
+    for (size_t k = 0; k < steps; ++k) {
+        RungeKutta4(*this, xEnd, s, ds);
+        // TODO can be improved by split integration.
+        calcStateProjectedToSurface(*this, xEnd);
+        // TODO don't log all points.
+        log.push_back(xEnd.postion);
+    }
+    std::cout << "s = " << s << std::endl;
+
+    return {std::move(xStart), std::move(xEnd)};
+}
+
+// TODO NOT YET IMPLEMENTED
+ImplicitSurfaceState
+ImplicitSurfaceParams::
+calcInitState(
+        const SimTK::Vec3& pStart,
+        const SimTK::Vec3& pEnd) const
+{
+    SimTK::Vec3 d = pEnd - pStart;
+    ImplicitSurfaceState q {pStart, d/d.norm()};
+    calcStateProjectedToSurface(*this, q);
+    return q;
+}
+
+// TODO NOT YET IMPLEMENTED
+double
+ImplicitSurfaceParams::
+calcInitLength(
+        const SimTK::Vec3& pStart,
+        const SimTK::Vec3& pEnd) const
+{
+    const SimTK::Vec3 pStartGuess = calcInitState(pStart, pEnd).postion;
+    const SimTK::Vec3 pEndGuess = calcInitState(pEnd, pStart).postion;
+    return (pStartGuess - pEndGuess).norm();
+}
+
+// TODO CORRECT?
+void
+ImplicitSurfaceParams::
+applyVariation(
+        const GeodesicVariation& var,
+        ImplicitSurfaceState& q) const
+{
+    DarbouxFrame frame = calcDarbouxFrame(*this, q);
+
+    q.postion += var.dBeta * frame.n
+        + var.dTangential * frame.t;
+    q.velocity += cos(var.dTheta) * frame.t
+        + sin(var.dTheta) * frame.b;
+
+    calcStateProjectedToSurface(*this, q);
+}
+
+//==============================================================================
+//                      IMPLICIT SPHERE IMPL
+//==============================================================================
```

Comment:
When we have multiple surfaces, it might be better to move this to an `ImplicitSurfaceImpl` file.